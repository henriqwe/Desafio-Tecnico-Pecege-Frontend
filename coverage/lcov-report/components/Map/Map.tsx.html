
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for components/Map/Map.tsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> / <a href="index.html">components/Map</a> Map.tsx</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">36.36% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>12/33</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">24% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>6/25</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">37.5% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>3/8</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">36.36% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>12/33</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a>
<a name='L44'></a><a href='#L44'>44</a>
<a name='L45'></a><a href='#L45'>45</a>
<a name='L46'></a><a href='#L46'>46</a>
<a name='L47'></a><a href='#L47'>47</a>
<a name='L48'></a><a href='#L48'>48</a>
<a name='L49'></a><a href='#L49'>49</a>
<a name='L50'></a><a href='#L50'>50</a>
<a name='L51'></a><a href='#L51'>51</a>
<a name='L52'></a><a href='#L52'>52</a>
<a name='L53'></a><a href='#L53'>53</a>
<a name='L54'></a><a href='#L54'>54</a>
<a name='L55'></a><a href='#L55'>55</a>
<a name='L56'></a><a href='#L56'>56</a>
<a name='L57'></a><a href='#L57'>57</a>
<a name='L58'></a><a href='#L58'>58</a>
<a name='L59'></a><a href='#L59'>59</a>
<a name='L60'></a><a href='#L60'>60</a>
<a name='L61'></a><a href='#L61'>61</a>
<a name='L62'></a><a href='#L62'>62</a>
<a name='L63'></a><a href='#L63'>63</a>
<a name='L64'></a><a href='#L64'>64</a>
<a name='L65'></a><a href='#L65'>65</a>
<a name='L66'></a><a href='#L66'>66</a>
<a name='L67'></a><a href='#L67'>67</a>
<a name='L68'></a><a href='#L68'>68</a>
<a name='L69'></a><a href='#L69'>69</a>
<a name='L70'></a><a href='#L70'>70</a>
<a name='L71'></a><a href='#L71'>71</a>
<a name='L72'></a><a href='#L72'>72</a>
<a name='L73'></a><a href='#L73'>73</a>
<a name='L74'></a><a href='#L74'>74</a>
<a name='L75'></a><a href='#L75'>75</a>
<a name='L76'></a><a href='#L76'>76</a>
<a name='L77'></a><a href='#L77'>77</a>
<a name='L78'></a><a href='#L78'>78</a>
<a name='L79'></a><a href='#L79'>79</a>
<a name='L80'></a><a href='#L80'>80</a>
<a name='L81'></a><a href='#L81'>81</a>
<a name='L82'></a><a href='#L82'>82</a>
<a name='L83'></a><a href='#L83'>83</a>
<a name='L84'></a><a href='#L84'>84</a>
<a name='L85'></a><a href='#L85'>85</a>
<a name='L86'></a><a href='#L86'>86</a>
<a name='L87'></a><a href='#L87'>87</a>
<a name='L88'></a><a href='#L88'>88</a>
<a name='L89'></a><a href='#L89'>89</a>
<a name='L90'></a><a href='#L90'>90</a>
<a name='L91'></a><a href='#L91'>91</a>
<a name='L92'></a><a href='#L92'>92</a>
<a name='L93'></a><a href='#L93'>93</a>
<a name='L94'></a><a href='#L94'>94</a>
<a name='L95'></a><a href='#L95'>95</a>
<a name='L96'></a><a href='#L96'>96</a>
<a name='L97'></a><a href='#L97'>97</a>
<a name='L98'></a><a href='#L98'>98</a>
<a name='L99'></a><a href='#L99'>99</a>
<a name='L100'></a><a href='#L100'>100</a>
<a name='L101'></a><a href='#L101'>101</a>
<a name='L102'></a><a href='#L102'>102</a>
<a name='L103'></a><a href='#L103'>103</a>
<a name='L104'></a><a href='#L104'>104</a>
<a name='L105'></a><a href='#L105'>105</a>
<a name='L106'></a><a href='#L106'>106</a>
<a name='L107'></a><a href='#L107'>107</a>
<a name='L108'></a><a href='#L108'>108</a>
<a name='L109'></a><a href='#L109'>109</a>
<a name='L110'></a><a href='#L110'>110</a>
<a name='L111'></a><a href='#L111'>111</a>
<a name='L112'></a><a href='#L112'>112</a>
<a name='L113'></a><a href='#L113'>113</a>
<a name='L114'></a><a href='#L114'>114</a>
<a name='L115'></a><a href='#L115'>115</a>
<a name='L116'></a><a href='#L116'>116</a>
<a name='L117'></a><a href='#L117'>117</a>
<a name='L118'></a><a href='#L118'>118</a>
<a name='L119'></a><a href='#L119'>119</a>
<a name='L120'></a><a href='#L120'>120</a>
<a name='L121'></a><a href='#L121'>121</a>
<a name='L122'></a><a href='#L122'>122</a>
<a name='L123'></a><a href='#L123'>123</a>
<a name='L124'></a><a href='#L124'>124</a>
<a name='L125'></a><a href='#L125'>125</a>
<a name='L126'></a><a href='#L126'>126</a>
<a name='L127'></a><a href='#L127'>127</a>
<a name='L128'></a><a href='#L128'>128</a>
<a name='L129'></a><a href='#L129'>129</a>
<a name='L130'></a><a href='#L130'>130</a>
<a name='L131'></a><a href='#L131'>131</a>
<a name='L132'></a><a href='#L132'>132</a>
<a name='L133'></a><a href='#L133'>133</a>
<a name='L134'></a><a href='#L134'>134</a>
<a name='L135'></a><a href='#L135'>135</a>
<a name='L136'></a><a href='#L136'>136</a>
<a name='L137'></a><a href='#L137'>137</a>
<a name='L138'></a><a href='#L138'>138</a></td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { memo, useCallback, useEffect, useState } from 'react'
import { GoogleMap, useJsApiLoader } from '@react-google-maps/api'
import type { IContact } from '&amp;types/contact'
import { MarkerDrop } from './MarkerDrop'
import { FormMarker } from './FormMarker'
&nbsp;
interface props {
  contacts: IContact[]
  showInfoWindow?: boolean
  onMapClick?: (e: google.maps.MapMouseEvent) =&gt; void
  formMarkerPosition?: { lat: number; lng: number }
}
&nbsp;
function MapComponent({
  contacts,
  showInfoWindow = true,
  onMapClick,
  formMarkerPosition
}: props) {
  const { isLoaded } = useJsApiLoader({
    id: 'google-map-script',
    googleMapsApiKey: process.env.VITE_GOOGLE_MAPS_API_KEY || ''
  })
&nbsp;
  const contactsWithLocation = contacts?.filter(
    (contact) =&gt; contact.address.geo.lat &amp;&amp; contact.address.geo.lng
  )
&nbsp;
  const [markers, setMarkers] = useState&lt;IContact[]&gt;([])
  const [formMarker, setFormMarker] = useState&lt;
    { lat: number; lng: number } | undefined
  &gt;()
&nbsp;
  const [_, setMap] = useState&lt;google.maps.Map | null&gt;(null)
&nbsp;
  function <span class="fstat-no" title="function not covered" >handleFitBounds(</span>map: google.maps.Map) {
    const bounds = <span class="cstat-no" title="statement not covered" >new window.google.maps.LatLngBounds()</span>
<span class="cstat-no" title="statement not covered" >    for (const contact of contactsWithLocation) {</span>
<span class="cstat-no" title="statement not covered" >      bounds.extend({</span>
        lat: Number(contact.address.geo.lat),
        lng: Number(contact.address.geo.lng)
      })
    }
<span class="cstat-no" title="statement not covered" >    map.fitBounds(bounds)</span>
  }
  const onLoad = useCallback(function <span class="fstat-no" title="function not covered" >callback(</span>map: google.maps.Map) {
    let generateFitBounds = <span class="cstat-no" title="statement not covered" >true</span>
&nbsp;
<span class="cstat-no" title="statement not covered" >    if (formMarkerPosition || contactsWithLocation.length === 1) {</span>
<span class="cstat-no" title="statement not covered" >      generateFitBounds = false</span>
&nbsp;
      const location = <span class="cstat-no" title="statement not covered" >{</span>
        lat:
          formMarkerPosition?.lat !== undefined
            ? formMarkerPosition?.lat
            : Number(contactsWithLocation[0].address.geo.lat),
        lng:
          formMarkerPosition?.lng !== undefined
            ? formMarkerPosition?.lng
            : Number(contactsWithLocation[0].address.geo.lng)
      }
<span class="cstat-no" title="statement not covered" >      map.setCenter(location)</span>
    }
<span class="cstat-no" title="statement not covered" >    if (contactsWithLocation.length &gt; 0) {</span>
<span class="cstat-no" title="statement not covered" >      setMarkers(contactsWithLocation)</span>
    }
&nbsp;
<span class="cstat-no" title="statement not covered" >    if (generateFitBounds) {</span>
<span class="cstat-no" title="statement not covered" >      handleFitBounds(map)</span>
    }
<span class="cstat-no" title="statement not covered" >    setMap(map)</span>
  }, [])
&nbsp;
  const onUnmount = useCallback(function <span class="fstat-no" title="function not covered" >callback(</span>) {
<span class="cstat-no" title="statement not covered" >    setMap(null)</span>
  }, [])
&nbsp;
  function <span class="fstat-no" title="function not covered" >handleMapClick(</span>event: google.maps.MapMouseEvent) {
<span class="cstat-no" title="statement not covered" >    if (onMapClick) {</span>
<span class="cstat-no" title="statement not covered" >      onMapClick(event)</span>
<span class="cstat-no" title="statement not covered" >      if (markers?.length &gt; 0) {</span>
<span class="cstat-no" title="statement not covered" >        setMarkers([])</span>
      }
<span class="cstat-no" title="statement not covered" >      setFormMarker({</span>
        lat: event.latLng?.lat() as number,
        lng: event.latLng?.lng() as number
      })
    }
  }
  useEffect(() =&gt; {
    setFormMarker(formMarkerPosition)
  }, [formMarkerPosition])
&nbsp;
  return isLoaded ? (
<span class="branch-0 cbranch-no" title="branch not covered" >    &lt;GoogleMap</span>
      mapContainerStyle={{
        width: '100%',
        height: '100%'
      }}
      onLoad={onLoad}
      onUnmount={onUnmount}
      // center={{
      //   lat: -12.100100128939063,
      //   lng: -49.24919742233473
      // }}
      data-testeid={'map'}
      zoom={3}
      options={{
        styles: [
          {
            featureType: 'poi',
            stylers: [{ visibility: 'off' }]
          },
          {
            featureType: 'transit',
            elementType: 'labels.icon',
            stylers: [{ visibility: 'off' }]
          }
        ]
      }}
      onClick={handleMapClick}
    &gt;
      {markers.map(<span class="fstat-no" title="function not covered" >(c</span>ontatc) =&gt; (
<span class="cstat-no" title="statement not covered" >        &lt;MarkerDrop</span>
          key={contatc.id}
          contact={contatc}
          showInfoWindow={showInfoWindow}
        /&gt;
      ))}
      {formMarker &amp;&amp; &lt;FormMarker position={formMarker} /&gt;}
    &lt;/GoogleMap&gt;
  ) : (
    &lt;&gt;&lt;/&gt;
  )
}
&nbsp;
export const Map = memo(MapComponent)
&nbsp;</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-08-20T21:24:57.016Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    