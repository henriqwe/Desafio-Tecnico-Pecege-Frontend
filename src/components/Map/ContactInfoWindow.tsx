import { Avatar } from '&components/Avatar/Avatar'
import { Button } from '&components/Button/Button'
import type { IContact } from '&types/contact'
import { routes } from '&utils/routes'
import { InfoWindow } from '@react-google-maps/api'
import { useNavigate } from 'react-router-dom'
import styled from 'styled-components'

interface props {
  contact: IContact
  onCloseClick: () => void
}

export function ContactInfoWindow({ contact, onCloseClick }: props) {
  const navigate = useNavigate()
  return (
    <InfoWindow
      key={'InfoWindow' + contact.id}
      onCloseClick={onCloseClick}
      position={{
        lat: Number(contact.address.geo.lat),
        lng: Number(contact.address.geo.lng)
      }}
    >
      <InfoWindowContent>
        <div className="grid-span-1 flex bg-gray-100  justify-center font-semibold rounded-l-md py-2 border-2 !border-white ">
          <Avatar name={contact.name} variant="sm" />
        </div>
        <ContatInfoWrapper>
          <PrimaryDetail>{contact.name}</PrimaryDetail>
          <SecondaryDetail>{contact.phone}</SecondaryDetail>
        </ContatInfoWrapper>
        <Button
          variant="secondary"
          type="button"
          onClick={() => navigate(routes.viewContact.path(contact.id))}
        >
          Ver contato
        </Button>
      </InfoWindowContent>
    </InfoWindow>
  )
}

const InfoWindowContent = styled.div`
  padding: 0.5rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.5rem;
`
export const PrimaryDetail = styled.span`
  font-size: 1.2rem;
  color: rgb(15 23 42);
  font-weight: 800;
`
export const SecondaryDetail = styled.span`
  font-size: 0.9rem;
  color: rgb(100 116 139);
  font-weight: 400;
`
export const ContatInfoWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`
