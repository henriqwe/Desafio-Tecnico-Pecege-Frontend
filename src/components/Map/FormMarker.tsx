import { Marker } from '@react-google-maps/api'

type props = {
  position: {
    lat: number
    lng: number
  }
}

export function FormMarker({ position }: props) {
  return (
    <Marker
      position={{
        lat: position.lat,
        lng: position.lng
      }}
      animation={google.maps.Animation?.DROP}
      zIndex={99}
    ></Marker>
  )
}
