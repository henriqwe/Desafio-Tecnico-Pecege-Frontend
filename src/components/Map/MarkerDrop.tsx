import { Marker } from '@react-google-maps/api'
import { useState } from 'react'
import { ContactInfoWindow } from './ContactInfoWindow'
import type { IContact } from '&types/contact'

type props = {
  contact: IContact
  showInfoWindow: boolean
}

export function MarkerDrop({ contact, showInfoWindow }: props) {
  const [openInfoWindow, setopenInfoWindow] = useState(false)

  function handleInfoWindow() {
    if (showInfoWindow) {
      setopenInfoWindow(!openInfoWindow)
    }
  }

  return (
    <Marker
      position={{
        lat: Number(contact.address.geo.lat),
        lng: Number(contact.address.geo.lng)
      }}
      animation={google.maps.Animation?.DROP}
      zIndex={99}
      data-testid={'marker'}
      onClick={handleInfoWindow}
    >
      {openInfoWindow && (
        <ContactInfoWindow contact={contact} onCloseClick={handleInfoWindow} />
      )}
    </Marker>
  )
}
