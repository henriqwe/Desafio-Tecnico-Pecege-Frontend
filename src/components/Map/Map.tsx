import { memo, useCallback, useEffect, useState } from 'react'
import { GoogleMap, useJsApiLoader } from '@react-google-maps/api'
import type { IContact } from '&types/contact'
import { MarkerDrop } from './MarkerDrop'
import { FormMarker } from './FormMarker'

interface props {
  contacts: IContact[]
  showInfoWindow?: boolean
  onMapClick?: (e: google.maps.MapMouseEvent) => void
  formMarkerPosition?: { lat: number; lng: number }
}

function MapComponent({
  contacts,
  showInfoWindow = true,
  onMapClick,
  formMarkerPosition
}: props) {
  const { isLoaded } = useJsApiLoader({
    id: 'google-map-script',
    googleMapsApiKey: process.env.VITE_GOOGLE_MAPS_API_KEY || ''
  })

  const contactsWithLocation = contacts?.filter(
    (contact) => contact.address.geo.lat && contact.address.geo.lng
  )

  const [markers, setMarkers] = useState<IContact[]>([])
  const [formMarker, setFormMarker] = useState<
    { lat: number; lng: number } | undefined
  >()

  const [map, setMap] = useState<google.maps.Map | null>(null)

  const onLoad = useCallback(function callback(map: google.maps.Map) {
    const bounds = new window.google.maps.LatLngBounds()
    if (formMarkerPosition) {
      bounds.extend({
        lat: Number(formMarkerPosition.lat),
        lng: Number(formMarkerPosition.lng)
      })
    }
    if (contactsWithLocation.length > 0) {
      for (const contact of contactsWithLocation) {
        bounds.extend({
          lat: Number(contact.address.geo.lat),
          lng: Number(contact.address.geo.lng)
        })
      }
      setMarkers(contactsWithLocation)
    }
    map.fitBounds(bounds)
    setMap(map)
  }, [])

  const onUnmount = useCallback(function callback(map: google.maps.Map) {
    setMap(null)
  }, [])

  function handleMapClick(event: google.maps.MapMouseEvent) {
    if (onMapClick) {
      onMapClick(event)
      if (markers?.length > 0) {
        setMarkers([])
      }
      setFormMarker({
        lat: event.latLng?.lat() as number,
        lng: event.latLng?.lng() as number
      })
    }
  }
  useEffect(() => {
    setFormMarker(formMarkerPosition)
  }, [formMarkerPosition])

  return isLoaded ? (
    <GoogleMap
      mapContainerStyle={{
        width: '100%',
        height: '100%'
      }}
      onLoad={onLoad}
      onUnmount={onUnmount}
      center={{
        lat: -12.100100128939063,
        lng: -49.24919742233473
      }}
      data-testeid={'map'}
      zoom={3}
      options={{
        styles: [
          {
            featureType: 'poi',
            stylers: [{ visibility: 'off' }]
          },
          {
            featureType: 'transit',
            elementType: 'labels.icon',
            stylers: [{ visibility: 'off' }]
          }
        ]
      }}
      onClick={handleMapClick}
    >
      {markers.map((contatc) => (
        <MarkerDrop
          key={contatc.id}
          contact={contatc}
          showInfoWindow={showInfoWindow}
        />
      ))}
      {formMarker && <FormMarker position={formMarker} />}
    </GoogleMap>
  ) : (
    <></>
  )
}

export const Map = memo(MapComponent)
