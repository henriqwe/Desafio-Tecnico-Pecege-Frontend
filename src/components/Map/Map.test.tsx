import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import { TestProviders } from '&utils/tests/TestProviders'
import { Map as MapContainer } from './Map'
import { initialize } from '@googlemaps/jest-mocks'

beforeEach(() => {
  initialize()
})

const contacts = [
  {
    id: 3,
    name: 'Clementine Bauch',
    username: 'Samantha',
    email: 'Nathan@yesenia.net',
    address: {
      street: 'Douglas Extension',
      suite: 'Suite 847',
      city: 'McKenziehaven',
      zipcode: '59590-4157',
      geo: {
        lat: '-68.6102',
        lng: '-47.0653'
      }
    },
    phone: '1-463-123-4447',
    website: 'ramiro.info',
    company: {
      name: 'Romaguera-Jacobson',
      catchPhrase: 'Face to face bifurcated interface',
      bs: 'e-enable strategic applications'
    }
  }
]
describe('EditContact', () => {
  it('should render the contact list correctly', async () => {
    const { container } = render(
      <TestProviders>
        <MapContainer contacts={contacts} />
      </TestProviders>
    )

    expect(container).toMatchSnapshot()
  })
})
