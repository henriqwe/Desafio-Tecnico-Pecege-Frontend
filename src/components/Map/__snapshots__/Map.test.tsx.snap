// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditContact should render the contact list correctly 1`] = `<div />`;
